### DISCUSSION TITLE
os.pullEvent for a peripheral

### DISCUSSION THREAD

**Discussion by CptToucan** [cc-tweaked/CC-Tweaked#1292] [Q&A]
I'm trying to use the redstoneIntegrator from Advanced Peripherals, and I want to use the os.pullEvent technique to wait for a redstone signal to be detected on the peripheral to let the code continue.

My problem is that I cannot work out how to get the code to run pullEvent for the peripheral. If this is not possible, what is my best course of action?

Thanks in advance.

EDIT: My understanding is that the redstoneIntegrator is only for outputting signals, if I wanted to wait for an event, the peripheral would need to implement its own event to wait for this, which is currently outside of the scope of what the integrator tries to do.

**Comment by fatboychummy** [comment]
I have asked SrEndi (the developer of Advanced Peripherals) on Discord if there is any event generated, he states there is not.

Unfortunately the only way to "listen" for updates is to have something that constantly polls all integrators for any changes.

![SmartSelect_20230106_094310_Discord](https://user-images.githubusercontent.com/39929480/211057772-36cd4c76-c173-476f-8e57-0bbbb529bd43.jpg)

Edit: SrEndi states that this is on his todo list

**Reply by fatboychummy** [reply]
I'll write something up here momentarily that could be used as a workaround, but I'm at work so time is sparse.

**Reply by fatboychummy** [reply]
```lua
-- Integrator read speed.
-- Note there are 10 redstone ticks per second, so 0.1 works here.
-- If you don't need to poll that quickly, increase this value.
local INTEGRATOR_TICK_SPEED = 0.1

-- information store
local integrator_data = {}
local sides = rs.getSides()

--- Check redstone integrators and queue events for the ones that have changed.
local function check_integrators()
  local integrators = table.pack(peripheral.find("redstoneIntegrator"))

  -- for each redstone integrator...
  for _, integrator in ipairs(integrators) do
    -- check that we have cached information for the integrator
    -- if not, create a blank cache for it.
    local name = peripheral.getName(integrator)
    if not integrator_data[name] then
      integrator_data[name] = {}
    end
    local cache = integrator_data[name] -- quick-reference variable
    
    -- check redstone inputs for the integrator
    for _, side in ipairs(sides) do
      local new = integrator.getAnalogInput(side)

      -- if the side data changed
      if new ~= cache[side] then
        -- generate an event:
        -- event data: integrator_change, integrator, side, old_value, new_value
        os.queueEvent("integrator_change", name, side, cache[side], new)

        -- store the new value
        cache[side] = new
      end
    end
  end
end

--- Loop the redstone integrator check function
local function run_integrators()
  while true do
    check_integrators()
    sleep(INTEGRATOR_TICK_SPEED)
  end
end

--- Your program here
local function main()
  -- paste your code here
end

parallel.waitForAny(run_integrators, main)
```

Here's a workaround. It's not great to have to do this, but it'll work. This will generate events with the signature of `"integrator_change", integrator_name, side_changed, old_value, new_value`.

**Reply by fatboychummy** [reply]
Please note that this was written on my phone, there may be slight errors I have missed!

**Reply by SirEndii** [reply]
> Edit: SrEndi states that this is on his todo list
 
 😸
